drivers/dac/ad5754r/ad5754r.c

```diff
@@ -0,0 +1,767 @@
+
+		if (bin_2scomp == NO_OS_GPIO_HIGH) {
+			code -= 1 << (AD5754R_MAX_RESOLUTION - 1);
+		} else {
+			if (code >= (1 << (AD5754R_MAX_RESOLUTION - 1))) {
+				code = (1 << AD5754R_MAX_RESOLUTION) - code;
+			}
+		}
+
+		*value = ad5754r_gain_values[dev->dac_ch_range[chn]]*vref*code /
+			 (1 << AD5754R_MAX_RESOLUTION);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Convert DAC millivolts to code for specific channel.
+ * @param dev - The device structure.
+ * @param chn - The index of channel being addressed.
+ * @param mvolts - Desired millivolts ouput for specific channel
+ * @param vref - Input reference voltage to AD5754R
+ * @param code - DAC code for desired output voltage
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5754r_dac_mvolts_to_code(struct ad5754r_dev *dev,
+			       uint8_t chn,
+			       uint16_t mvolts,
+			       int vref,
+			       uint16_t *code)
+{
+	int ret;
+	uint8_t bin_2scomp;
+
+	if (!dev || !code || (chn >= AD5754R_NUM_CHANNELS))
+		return -EINVAL;
+
+	/* Based on unipolar/bipolar output range, calculate code */
+	if (dev->dac_ch_range[chn] < AD5754R_SPAN_M5V_TO_5V) {
+		*code = ((1 << AD5754R_MAX_RESOLUTION) * mvolts) /
+			(vref * ad5754r_gain_values[dev->dac_ch_range[chn]]);
+	} else {
+		if (!dev->gpio_bin_2scomp) {
+			return -EINVAL; // No way to tell output coding
+		}
+
+		ret = no_os_gpio_get_value(dev->gpio_bin_2scomp, &bin_2scomp);
+		if (ret)
+			return ret;
+
+		*code = (1 << AD5754R_MAX_RESOLUTION) * mvolts/
+			(ad5754r_gain_values[dev->dac_ch_range[chn]]*vref);
+
+		if (bin_2scomp == NO_OS_GPIO_HIGH) {
+			*code += 1 << (AD5754R_MAX_RESOLUTION - 1);
+		} else {
+			if (*code >= (1 << (AD5754R_MAX_RESOLUTION - 1))) {
+				*code -= 1 << AD5754R_MAX_RESOLUTION ;
+			}
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Set output range for specific channel.
+ * @param dev - The device structure.
+ * @param chn - The index of channel being addressed.
+ * @param ch_range - Range option to be set for specified channel.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5754r_set_ch_range(struct ad5754r_dev *dev, uint8_t chn,
+			 enum ad5754r_dac_ch_range ch_range)
+{
+	int ret;
+
+	if ((chn > AD5754R_NUM_CHANNELS)|| (ch_range > AD5754R_SPAN_M10V8_TO_10V8))
+		return -EINVAL;
+
+	ret = ad5754r_write(dev,
+			    AD5754R_PREP_INSTR_ADDR(AD5754R_REG_OUTPUT_RANGE_SEL, chn),
+			    ch_range);
+	if (ret)
+		return ret;
+
+	dev->dac_ch_range[chn] = ch_range;
+
+	return 0;
+}
+
+/**
+ * @brief Set power-up state for specific channel.
+ * @param dev - The device structure.
+ * @param chn - The index of channel being addressed.
+ * @param ch_pwrup - Powerup setting for specified channel.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5754r_set_ch_pwrup(struct ad5754r_dev *dev, uint8_t chn,
+			 enum ad5754r_pwr_dac_ch_pwrup ch_pwrup)
+{
+	int ret;
+
+	if ((chn >= AD5754R_NUM_CHANNELS) ||
+	    (ch_pwrup > AD5754R_PWR_DAC_CH_POWERUP))
```

Comment:
Since there are only 2 options for pwr_up enum, it might be a better idea to just have this as boolean input (TRUE or FALSE only).  Doing so will eliminate the need to check if the ch_pwrup is within range.